### ISSUE TITLE
New Snowflake `group by all` unparseable by SQLFluff

### ISSUE THREAD

**Issue by MartyC-137** [sqlfluff/sqlfluff#4940]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

The new Snowflake [`group by all`](https://docs.snowflake.com/en/sql-reference/constructs/group-by) functionality is unparseable by SQLFluff. Note that I am willing to work on a PR but may need some guidance on what needs to be modified.

### Expected Behaviour

This should be parseable for SQLFluff as it is now valid syntax for Snowflake.

### Observed Behaviour

```bash
== [group_by_all_repro.sql] FAIL
L:   6 | P:   1 |  PRS | Line 6, Position 1: Found unparsable section: 'group by
                       | all'
WARNING: Parsing errors found and dialect is set to 'snowflake'. Have you configured your dialect correctly?
All Finished 📜 🎉!
```

### How to reproduce

Set up the example in **Snowflake**:

```sql
-- Change these values accordingly
set db = 'my_db';
set sch = 'my_schema';
set table_name = 'sales';
set role_name = 'my_role';
set wh = 'my_wh';

use warehouse identifier($wh);
use role identifier($role_name);
use database identifier($db);
use schema identifier($sch);

create table identifier($table_name) (
  product_ID integer,
  retail_price real,
  quantity integer,
  city varchar,
  state varchar);

insert into sales (product_id, retail_price, quantity, city, state) values
  (1, 2.00,  1, 'SF', 'CA'),
  (1, 2.00,  2, 'SJ', 'CA'),
  (2, 5.00,  4, 'SF', 'CA'),
  (2, 5.00,  8, 'SJ', 'CA'),
  (2, 5.00, 16, 'Miami', 'FL'),
  (2, 5.00, 32, 'Orlando', 'FL'),
  (2, 5.00, 64, 'SJ', 'PR');
```

Create a file called `group_by_all_repro.sql` with the following content:

```sql
select
    state,
    city,
    sum(retail_price * quantity) as gross_revenue
from sales
group by all;

```

Run `sqlfluff lint` on the file:

```bash
$ sqlfluff lint group_by_all_repro.sql --dialect snowflake
```

Error:

```bash
$ sqlfluff lint group_by_all_repro.sql --dialect snowflake
== [group_by_all_repro.sql] FAIL
L:   6 | P:   1 |  PRS | Line 6, Position 1: Found unparsable section: 'group by
                       | all'
WARNING: Parsing errors found and dialect is set to 'snowflake'. Have you configured your dialect correctly?
All Finished 📜 🎉!
```

### Dialect

Snowflake

### Version

2.1.1

### Configuration

```ini
[sqlfluff]
templater = jinja
dialect = snowflake
warnings = LT05,RF02,AL05
exclude_rules = RF01
max_line_length = 150
ignore = templating

[sqlfluff:templater:jinja]
load_macros_from_path = macros

[sqlfluff:indentation]
allow_implicit_indents = true

[sqlfluff:rules:aliasing.table]
aliasing = explicit

[sqlfluff:rules:aliasing.column]
aliasing = explicit

[sqlfluff:rules:aliasing.length]
min_alias_length = 3

[sqlfluff:rules:layout.long_lines]
ignore_comment_clauses = True
ignore_comment_lines = True

[sqlfluff:rules:capitalisation.keywords]
capitalisation_policy = lower

[sqlfluff:rules:capitalisation.identifiers]
capitalisation_policy = lower

[sqlfluff:rules:capitalisation.functions]
extended_capitalisation_policy = lower

[sqlfluff:rules:capitalisation.literals]
capitalisation_policy = lower

[sqlfluff:rules:capitalisation.types]
extended_capitalisation_policy = lower

[sqlfluff:rules:references.special_chars]
quoted_identifiers_policy = None
```

### Are you willing to work on and submit a PR to address the issue?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by MartyC-137** [comment]
I think `class GroupByClauseSegment` in [`sqlfluff/src/dialects/snowflake/dialect_snowflake.py`](https://github.com/sqlfluff/sqlfluff/blob/main/src/sqlfluff/dialects/dialect_snowflake.py#L822) needs to be updated to include `all` in the Sequence:

```py
class GroupByClauseSegment(ansi.GroupByClauseSegment):
    """A `GROUP BY` clause like in `SELECT`.

    Snowflake supports Cube, Rollup, and Grouping Sets

    https://docs.snowflake.com/en/sql-reference/constructs/group-by.html
    """

    match_grammar: Matchable = Sequence(
        "GROUP",
        "BY",
        Indent,
        OneOf(
            Sequence(
                OneOf("CUBE", "ROLLUP", "ALL", Sequence("GROUPING", "SETS")),
                Bracketed(
                    Ref("GroupByContentsGrammar"),
                ),
            ),
            Ref("GroupByContentsGrammar"),
        ),
        Dedent,
    )

```

Looks like I would also need to add a `.sql` file with a basic example and an autogenerated `.yml` file for that example using `test/generate_parse_fixture_yml.py`? If someone can confirm that this seems like the correct approach, I can work the PR for this.

**Comment by greg-finley** [comment]
@MartyC-137 yep, that sounds like a good approach 👍

**Comment by Sunnyinho** [comment]
Thanks for raising this issue @MartyC-137.

**Comment by Sunnyinho** [comment]
Does this(group by all) work in the latest version 2.1.2 now? Because I upgraded  my sqlfluff to 2.1.2 form 2.1.0 and it still throws me the error 
```bash
L:  59 | P:   5 |  PRS | Line 59, Position 5: Found unparsable section: 'group by
                       | all'
WARNING: Parsing errors found and dialect is set to 'snowflake'. Have you configured your dialect correctly?
```

**Comment by greg-finley** [comment]
Nope, it’s not released yet.

**Comment by jamiekt** [comment]
Hello. Which version of sqlfluff did this fix first appear in?
